### DISCUSSION TITLE
Endpoint creation function helper

### DISCUSSION THREAD

**Discussion by Zehir** [ecyrbe/zodios#420] [Q&A]
Hello,

I am trying to create an helper function that create the endpoint because the api are all similar; 

Like this method :
https://github.com/ecyrbe/zodios/blob/9c6686d5a739e752635894feec9979d785cb2b8e/src/api.ts#L237-L240

Here what I made, it's just a simpler version but even this one I can't make it work properly.

```typescript
interface CustomZodiosEndpointDefinition<D extends z.ZodType<any> = z.ZodType<any>>
  extends Pick<ZodiosEndpointDefinition, 'alias' | 'description' | 'method' | 'path' | 'parameters'> {
  response: D
  unwrapResponse?: boolean
  body?: z.ZodType<any>
  errorCodes?: number[]
}

export function makeCustomEndpoint<T extends CustomZodiosEndpointDefinition>(endpoint: T) {
  return makeEndpoint({
    alias: endpoint.alias,
    description: endpoint.description,
    method: endpoint.method,
    path: endpoint.path,
    parameters: endpoint.parameters,
    response: (endpoint.unwrapResponse === true
      ? unwrap(endpoint.path, endpoint.response)
      : endpoint.response),
  })
}
```

The method do works but the typedefinition is messedup.
And when I try to use it with :
```typescript

  makeEndpoint({
    alias: 'createAPIKey',
    description: 'Creates a new API key which provides authorized access to the REST-API. '
    + 'The request will only succeed if the gateway is unlocked or valid HTTP basic '
    + 'authentification credentials are provided in the HTTP request header see authorization.',
    method: 'post',
    path: '/api',
    response: unwrap('/api', z.strictObject({ username: z.string() }).transform(result => result.username)),
  }),

  makeCustomEndpoint({
    alias: 'createAPIKeyTest',
    description: 'Creates a new API key which provides authorized access to the REST-API. '
    + 'The request will only succeed if the gateway is unlocked or valid HTTP basic '
    + 'authentification credentials are provided in the HTTP request header see authorization.',
    method: 'post',
    path: '/api/',
    unwrapResponse: true,
    response: z.strictObject({ username: z.string() }).transform(result => result.username),
  }),
```

For the first one I get a perfect valid typescript definition but for the helper method I get :
```typescript
{
	alias: string | undefined;
	description: string | undefined;
	method: import("@zodios/core/lib/zodios.types").Method;
	path: string;
	parameters: import("@zodios/core/lib/zodios.types").ZodiosEndpointParameter<unknown>[] | undefined;
	response: import("zod").ZodType<any, import("zod").ZodTypeDef, any>;
}
```

Thanks for your help

**Comment by ecyrbe** [comment]
converting to discussion as this is not an issue with zodios.

**Reply by Zehir** [reply]
Sorry I forgot that discussions was a thing

**Reply by davedc** [reply]
@Zehir were you able to solve this? I ran into something similar.

**Reply by Zehir** [reply]
> @Zehir were you able to solve this? I ran into something similar. 
> 
> 

No I rage quit ;)

**Reply by ecyrbe** [reply]
In general, making these helpers require you to do two things :
- make a runtime transform impl
- make a compile time type transform impl

If you are not versed in advanced Typescript type meta programming, i don't recommand trying to do so. You'll loose more Time doing them than what you'll gain. And even worse, your team might not be able to understand and maintain what you did.

Zodios relying heavilly on custom utility types, i don't see a lot of PR to add features because not a lot of users understand how to do them.

So on user code, i recommand you to keep it simple and use the native zodios format. I maintain it for you.

**Comment by dejay-at-upstart** [comment]
Using Zodios, it seems really hard to avoid needing object literals defined and duplicated all over one's code base.  The purpose of utility functions and custom builders are to centralize and simplify.  I haven't found any satisfying way of doing this with Zodios.

**Reply by ecyrbe** [reply]
use aliases, they are there for that.
